player/js/utils/PolynomialBezier.js

```diff
@@ -37,6 +55,41 @@ function polynomialCoefficients(p0, p1, p2, p3) {
   ];
 }
 
+/**
+ * Returns the _normalized_ direction vector of the tangent of the curve when approaching
+ * the point from negative infinity, pointing outwards at cusps.
+ */
+function tangentDirection(p0, to, ti, p3, t) {
+  const controlPoints = [p0, addVectors(p0, to), addVectors(p3, ti), p3];
+  const firstSubdivision = deCasteljauSubdivision(controlPoints, t);
+  const secondSubdivision = deCasteljauSubdivision(firstSubdivision, t);
+
+  const tangent1 = subtractVectors(secondSubdivision[1], secondSubdivision[0]);
+  const length1 = Math.hypot(...tangent1);
+  if (length1 > Number.EPSILON) return tangent1.map((v) => v / length1);
+
+  const tangent2 = subtractVectors(firstSubdivision[2], firstSubdivision[0]);
+  const length2 = Math.hypot(tangent2);
```

Comment:
Is this supposed to be `Math.hypot(...tangent2)`?

Given that we don't have the normal TypeScript handrail here, and no unit testing infra to speak of, can I trouble you to please verify the three branches here (L73, L79, L81) with a manual test?

You can use the browser console, a Node environment, whatever is easiest.

Reply:
Sure thing.
- `document.tangentDirection([-8, -4], [2, 6], [-2, 6], [8, -4], 0.5)`, should be `[1, 0]`, for the first branch
- `document.tangentDirection([-8, -4], [0, 0], [-2, 6], [8, -4], 0) `, should be `[0.91, 0.39]` for the second branch (stationary point)
- `document.tangentDirection([-8, -4], [16, 6], [-16, 6], [8, -4], 0.5)`, should be `[0, 1]` for the third branch (cusp)
- `document.tangentDirection([-8, -4], [16, 0], [0, 0], [8, -4], 1) `, should be `[1, 0]` for the fourth branch (curve is a line)
- `document.tangentDirection([0, 0], [0, 0], [0, 0], [0, 0], 1) `, should be `[0, 0]` for the fourth branch (degenerate curve)

Reply:
Also, sorry for the typo. I'll try to pay more attention.

Reply:
Nice! Thank you so much!